vyper/venom/analysis/available_expression.py

```diff
@@ -0,0 +1,363 @@
+from __future__ import annotations
+
+from collections import deque
+from dataclasses import dataclass
+from functools import cached_property
+
+import vyper.venom.effects as effects
+from vyper.venom.analysis.analysis import IRAnalysesCache, IRAnalysis
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.analysis.dfg import DFGAnalysis
+from vyper.venom.basicblock import (
+    BB_TERMINATORS,
+    COMMUTATIVE_INSTRUCTIONS,
+    IRBasicBlock,
+    IRInstruction,
+    IROperand,
+    IRVariable,
+)
+from vyper.venom.context import IRFunction
+from vyper.venom.effects import Effects
+
+NONIDEMPOTENT_INSTRUCTIONS = frozenset(["log", "call", "staticcall", "delegatecall", "invoke"])
+
+# instructions that queries info about current
+# environment this is done because we know that
+# all these instruction should have always
+# the same value in function
+
+
+# instruction that dont need to be stored in available expression
+UNINTERESTING_OPCODES = frozenset(
+    [
+        "calldatasize",
+        "gaslimit",
+        "address",
+        "codesize",
+        "store",
+        "phi",
+        "param",
+        "nop",
+        "returndatasize",
+        "gas",
+        "gasprice",
+        "origin",
+        "coinbase",
+        "timestamp",
+        "number",
+        "prevrandao",
+        "chainid",
+        "basefee",
+        "blobbasefee",
+        "pc",
+        "msize",
+    ]
+)
+
+
+@dataclass
+class _Expression:
+    opcode: str
+    # the child is either expression of operand since
+    # there are possibilities for cycles
+    operands: list[IROperand | _Expression]
+    ignore_msize: bool
```

Comment:
i think `ignore_msize` does not really belong inside the `_Expression` class. and the way we are using it, it can be handled in the caller (of `get_reads()` and `get_writes()`)